bpf/tests/skip_lb_xlate_lrp_per_packet_lb.c

```diff
@@ -0,0 +1,333 @@
+	/* Push IPv4 header */
+	l3 = pktgen__push_default_iphdr(&builder);
+	if (!l3)
+		return TEST_ERROR;
+	l3->saddr = V4_BACKEND_IP;
+	l3->daddr = V4_SERVICE_IP;
+	/* Push TCP header */
+	l4 = pktgen__push_default_tcphdr(&builder);
+	if (!l4)
+		return TEST_ERROR;
+	l4->source = tcp_src_one;
+	l4->dest = SERVICE_PORT;
+	data = pktgen__push_data(&builder, default_data, sizeof(default_data));
+	if (!data)
+		return TEST_ERROR;
+	/* Calc lengths, set protocol fields and calc checksums */
+	pktgen__finish(&builder);
+	return 0;
+}
+
+SETUP("tc", "v4_local_redirect")
+int v4_local_backend_to_service_setup(struct __ctx_buff *ctx)
+{
+	__u16 revnat_id = 1;
+	struct lb4_key lb_svc_key = {};
+	struct lb4_service lb_svc_value = {};
+	struct lb4_reverse_nat revnat_value = {};
+	struct lb4_backend backend = {};
+	struct ipcache_key cache_key = {};
+	struct remote_endpoint_info cache_value = {};
+	struct endpoint_key ep_key = {};
+	struct endpoint_info ep_value = {};
+	/* Register a fake LB backend for our local redirect service */
+	lb_svc_key.address = V4_SERVICE_IP;
+	lb_svc_key.dport = SERVICE_PORT;
+	lb_svc_key.scope = LB_LOOKUP_SCOPE_EXT;
+	/* Create a LRP service with one backend */
+	lb_svc_value.count = 1;
+	lb_svc_value.flags = SVC_FLAG_ROUTABLE;
+	lb_svc_value.flags2 = SVC_FLAG_LOCALREDIRECT;
+	lb_svc_value.rev_nat_index = revnat_id;
+	map_update_elem(&LB4_SERVICES_MAP_V2, &lb_svc_key, &lb_svc_value, BPF_ANY);
+	/* Insert a reverse NAT entry for the above service */
+	revnat_value.address = V4_SERVICE_IP;
+	revnat_value.port = SERVICE_PORT;
+	map_update_elem(&LB4_REVERSE_NAT_MAP, &revnat_id, &revnat_value, BPF_ANY);
+	/* A backend between 1 and .count is chosen, since we have only one backend
+	 * it is always backend_slot 1. Point it to backend_id 124.
+	 */
+	lb_svc_key.backend_slot = 1;
+	lb_svc_value.backend_id = 124;
+	map_update_elem(&LB4_SERVICES_MAP_V2, &lb_svc_key, &lb_svc_value, BPF_ANY);
+
+	/* Add the service in LB4_SKIP_MAP to skip service translation for request originating from the local backend */
+	struct skip_lb4_key key = {
+		.netns_cookie = ENDPOINT_NETNS_COOKIE,
+		.address = V4_SERVICE_IP,
+		.port = SERVICE_PORT,
+	};
+	__u8 val = 0;
+
+	map_update_elem(&LB4_SKIP_MAP, &key, &val, BPF_ANY);
+
+	/* Create backend id 124 which contains the IP and port to send the
+	 * packet to.
+	 */
+	backend.address = V4_BACKEND_IP;
+	backend.port = BACKEND_PORT;
+	backend.proto = IPPROTO_TCP;
+	backend.flags = 0;
+	map_update_elem(&LB4_BACKEND_MAP, &lb_svc_value.backend_id, &backend, BPF_ANY);
+	/* Add an IPCache entry for pod 1 */
+	cache_key.lpm_key.prefixlen = 32;
+	cache_key.family = ENDPOINT_KEY_IPV4;
+	cache_key.ip4 = V4_BACKEND_IP;
+	/* a random sec id for the pod */
+	cache_value.sec_identity = 112233;
+	map_update_elem(&IPCACHE_MAP, &cache_key, &cache_value, BPF_ANY);
+	ep_key.ip4 = V4_BACKEND_IP;
+	ep_key.family = ENDPOINT_KEY_IPV4;
+	map_update_elem(&ENDPOINTS_MAP, &ep_key, &ep_value, BPF_ANY);
```

Comment:
I want to believe that we have helpers for most of this in `bpf/tests/lib/...` now. Except the `LB4_SKIP_MAP` operations, which would be a good addition?

Reply:
Thanks. Created https://github.com/cilium/cilium/pull/34443.